## ***Manipulação com strings***
---

uma forma de centralizar ou colocar nas extremidades do output

In [1]:
print(f'{"a":^5}.')

  a  .


outra forma de centralizar é o método `center()`, que recebe como parâmtro a quantidade de caracteres que a string deve ser centralizada e quais caracteres preencher esse espaço

In [20]:
print(f'{"a".center(5)}.')

  a  .


In [21]:
print(f'{"a".center(5, "*")}')

**a**


In [2]:
print(f'{"a":<5}.')

a    .


pode usar o método `ljust()` da mesma forma que `center()` para posicionar a string à esquerda

In [3]:
print(f'{"a":>5}.')

    a.


pode usar o método `rjust()` da mesma forma que `center()` para posicionar a string à direita

os espaços podem ser preenchidos com outros caracteres que não só espaços:

In [4]:
print(f'{"a":*^5}.')

**a**.


In [5]:
print(f'{"a":b^5}.')

bbabb.


o operador `\t` serve para criar uma formatação de tabela:

In [1]:
print('\ta')

	a


números também podem ser formatados usando essas mesmas formas, com exceção dos métodos

além de que pode-se usar vírgula para separar casas de milha

In [28]:
print(f'{7532:,}')

7,532


e pode ser usado junto das casas decimais

In [29]:
print(f'{83442320/100:,.5f}')

834,423.20000


bem como pode-se tranformar números em outras bases

In [32]:
print(f'{45:0b}')

101101


a tabela mostra todas as transformações de base possível pelo python

COD | DESCRIÇÃO | p.ex.: (45)
---|---|---
0b | binário | 101101
0c | caractere | -
0d | base 10 | 45
0h | base 10 local | 45
0o | octal | 55
x | hexadecimal minúsculo | 2d
X | hexadecimal maiúsculo | 2D

para formatação de números decimais

COD | DESCRIÇÃO | p.ex.: (1.345)
---|---|---
g | genérico | 1.345
G | genérico | 1.345
n | local | 1,345
% | percentual | 134.500000%

#### Análises
---

o método ```count()``` recebe uma string como parâmetro e retorna a quantidade de vezes que essa string aparece

In [33]:
stg = 'hello, world!'
print(stg.count('l'))

3


```count()``` pode, ainda, recebe um intervalo após a string, e em ```stg``` ele buscará apenas dentre desse intervalo

In [8]:
print(stg.count('l', 0, 5))
print(stg[0:5])

2
hello


o método ```find()``` recebe uma string e verifica em qual posição essa string aparece

In [9]:
print(stg.find('rl'))

9


`find()` ainda pode receber um intervalo na qual deve ser procurada a string

In [10]:
print(stg.find('rl', 6, 11))

9


usando `rfind()` buscará a partir da direita da string

o problema de usar `find()` é que, se a string se repete dentro da variável, ele erá contar apenas a 1ª aparição dessa string.

nesse caso, pode ser usado o seguinte código para encontrar todas ocorrências

In [17]:
plv = 'um tigre, dois tigres, três tigres'
p = 0
while p>-1:
    p = plv.find('tigre', p)
    if p>=0:
        print(f'posição: {p}')
        p+=1

posição: 3
posição: 15
posição: 28


em contra-partida do método `find()` ou `rfind()`

In [18]:
print(f'posição: {plv.find("tigre")}')

posição: 3


In [19]:
print(f'posição: {plv.rfind("tigre")}')

posição: 28


#### Transformações
---

strings são imutáveis, porém, podem ser manipuladas  das seguintes formas

o método `replace()` recebe duas string como parâmetro: a 1ª é a parte da string que deseja-se mudar e a 2ª é a para o que será mudado; pode ainda receber quantas strings deseja-se mudar

In [11]:
print(stg.replace('hello', 'goodbye'))

goodbye, world!


In [25]:
print(plv.replace('tigre', 'gato', 2))

um gato, dois gatos, três tigres


se a primeira string for vazia, o método `replace()` incluirá a segunda string entre todas os caracteres que formam a variável

In [26]:
print(plv.replace('', '*'))

*u*m* *t*i*g*r*e*,* *d*o*i*s* *t*i*g*r*e*s*,* *t*r*ê*s* *t*i*g*r*e*s*


bem como pode ser apagada a primeira string se for passada uma string vazia na segunda

In [27]:
print(plv.replace('tigre', ''))

um , dois s, três s


o método ```split()``` não necessariamente recebe parâmetros, mas pode receber um caractere que será usado para separar cada string em uma lista

In [12]:
print(stg.split())
print(stg.split('l'))

['hello,', 'world!']
['he', '', 'o, wor', 'd!']


e o ```'<caractere_de_junção>'.join(variável)``` recebe um caractere_de_junção que irá juntar a lista separada passada como parâmetro 

In [13]:
print('*'.join(stg))

h*e*l*l*o*,* *w*o*r*l*d*!


para separar cada caractere da string em um elemento diferente em uma lista, faça:

In [14]:
print(list(stg))

['h', 'e', 'l', 'l', 'o', ',', ' ', 'w', 'o', 'r', 'l', 'd', '!']


além dos já conhecidos `upper()`, `lower()`, `capitalize()`, `title()` e `strip()`.
os métodos `rstrip()` e `lstrip(` apagam espaços extras sem funcionalidades no lado direito ou esquerdo da string, respectivamente.

#### Verificação
---

pode-se usa os métodos `endswith()`, `startswith()` que recebem strings como parâmetros e verificam se a variável começa ou não com a string passada.

os métodos `isalnum()`, `isalpha()`, `isascii()`, `isdecimal()`, `isdigit()`, `islower()`, `isprintable()`, `isnumeric()`, `isspace()`, `istitle()`, `isupper()` não recebem nenhum parâmetro, mas retornam o que o próprio nome dos métodos indicam

In [15]:
print(stg.endswith('!'))
print(stg.startswith('m'))
print(stg.isalnum())
print(stg.isalpha())
print(stg.isascii())
print(stg.isdecimal())
print(stg.isdigit())
print(stg.islower())
print(stg.isprintable())
print(stg.isnumeric())
print(stg.isspace())
print(stg.istitle())
print(stg.isupper())

True
False
False
False
True
False
False
True
True
False
False
False
False


#### comparação
---

uma pode ser comparada se vem antes no alfabeto do que uma outra, ou vice-versa, usando os operadores `<` e `>`

In [1]:
print('lucas' > 'matheus')
print('lucas' < 'matheus')

False
True


pode, ainda, usar `==`, `!=` para verificaar se duas strings são iguais ou não:

In [2]:
print('lucas' == 'matheus')
print('lucas' != 'matheus')

False
True


#### transformando string em números
---

para é necessário usar a função `ord()`

In [7]:
print(ord('w'))

119


para reverter, ou transformar um número qualquer em string, podemos usar `chr()`

In [8]:
print(chr(119))

w


#### enchergando espaços
---

a função integrada `repr(<string>)` pode ser usada para mostrar os espaços em branco presente em uma string:

In [1]:
stg = '1 2\t3\n4'
print(stg)

1 2	3
4


In [2]:
print(repr(stg))

'1 2\t3\n4'
